indra/llcommon/llcallbacklist.h

```diff
@@ -27,53 +27,248 @@
-
 	callback_list_t	mCallbackList;
+
+	// "Additional specializations for std::pair and the standard container
+	// types, as well as utility functions to compose hashes are available in
+	// boost::hash."
+	// https://en.cppreference.com/w/cpp/utility/hash
+	typedef std::pair< callback_t,void* >	callback_pair_t;
+	typedef std::unordered_map<callback_pair_t, handle_t,
+		boost::hash<callback_pair_t>> lookup_table;
+	lookup_table mLookup;
 };
 
-typedef boost::function<void ()> nullary_func_t;
-typedef boost::function<bool ()> bool_func_t;
+/*-------------------- legacy names in global namespace --------------------*/
+#define gIdleCallbacks (LLCallbackList::instance())
+
+using nullary_func_t = LLCallbackList::callable_t;
+using bool_func_t    = LLCallbackList::bool_func_t;
 
 // Call a given callable once in idle loop.
-void doOnIdleOneTime(nullary_func_t callable);
+inline
+LLCallbackList::handle_t doOnIdleOneTime(nullary_func_t callable)
+{
+    return gIdleCallbacks.doOnIdleOneTime(callable);
+}
 
 // Repeatedly call a callable in idle loop until it returns true.
-void doOnIdleRepeating(bool_func_t callable);
+inline
+LLCallbackList::handle_t doOnIdleRepeating(bool_func_t callable)
+{
+    return gIdleCallbacks.doOnIdleRepeating(callable);
+}
+
+/*****************************************************************************
+*   LLLater: callbacks at some future time
+*****************************************************************************/
+class LLLater: public LLSingleton<LLLater>
+{
+    LLSINGLETON(LLLater);
+
+    using token_t = U32;
+
+    // Define a struct for our priority queue entries, instead of using
+    // a tuple, because we need to define the comparison operator.
+    struct func_at
+    {
+        nullary_func_t mFunc;
+        token_t mToken;
+        LLDate::timestamp mTime;
+
+        func_at(const nullary_func_t& func, token_t token, LLDate::timestamp tm):
+            mFunc(func),
+            mToken(token),
+            mTime(tm)
+        {}
+
+        friend bool operator<(const func_at& lhs, const func_at& rhs)
+        {
+            // use greater-than because we want fibonacci_heap to select the
+            // EARLIEST time as the top()
+            return lhs.mTime > rhs.mTime;
+        }
+    };
+
+    // Accept default stable<false>: when two funcs have the same timestamp,
+    // we don't care in what order they're called.
+    // Specify constant_time_size<false>: we don't need to optimize the size()
+    // method, iow we don't need to store and maintain a count of entries.
+    typedef boost::heap::fibonacci_heap<func_at, boost::heap::constant_time_size<false>>
+        queue_t;
+
+public:
+    // If tasks that come ready during a given tick() take longer than this,
+    // defer any subsequent ready tasks to a future tick() call.
+    static constexpr F32 TIMESLICE{ 0.005f };
+
+    class handle_t
+    {
+    private:
+        friend class LLLater;
```

Comment:
Friends are good to have in life. But why does handle_t need friendship and not a strict business relationship of having to call `handle_t( value )` and a new `token_t getToken() const` method?

Reply:
The intention of `handle_t` is to be completely opaque to consumers of `LLLater`. `token_t` is meant to be an internal-only implementation detail, because it might change.